File: SEGUNDO/C1/NetBeansProjects/Deep_space/src/deepspace/Damage.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;
/**
 *
 * @author daniel
 */
public class Damage {
    private int nShields;
    private int nWeapons;
    private ArrayList<WeaponType> weapons;
    
    public Damage(int w, int s){
        this.nShields=s;
        this.nWeapons=w;
    }
    
    public Damage(ArrayList<WeaponType> wl, int s){
        this.nShields=s;
        this.nWeapons=-1;
        this.weapons=(ArrayList<WeaponType>) wl.clone();
    }
    
    public Damage(Damage d){
        this.nShields=d.getNShields();
        this.nWeapons=d.getNWeapons();
        this.weapons=(ArrayList<WeaponType>) d.getWeapons().clone();
    }
    
    public DamageToUI getUIversion(){
        return new DamageToUI(this);
    }
    
    private int arrayContainsType(ArrayList<Weapon> w, WeaponType t){
        boolean encontrado=false;
        int posicion=0;
        
        for(int i = 0; i < w.size() && !encontrado; i++){
            if(w.get(i).getType() == t){
                encontrado=true;
                posicion=i;
            }
                
        }
        
        if(!encontrado)
            posicion=-1;
        
        return posicion;
    }
    
    public Damage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s){
        if(weapons.isEmpty()){
            nShields=Math.min(nShields,s.size());
            nWeapons=Math.min(nWeapons, w.size());
            
            return new Damage(nWeapons,nShields);
        }
        
        else{
            nShields=Math.min(nShields, s.size());
            
            int lasers_1=0, lasers_2=0;
            int plasmas_1=0, plasmas_2=0;
            int misiles_1=0,misiles_2=0;
            
            for(int i=0; i < w.size(); i++){
                if(w.get(i).getType() == WeaponType.LASER)
                    lasers_1++;
                
                else if(w.get(i).getType() == WeaponType.PLASMA)
                    plasmas_1++;
                
                else
                    misiles_1++;
            }
            
            for(int i=0; i < weapons.size(); i++){
                if(weapons.get(i) == WeaponType.LASER)
                    lasers_2++;
                
                else if(weapons.get(i) == WeaponType.PLASMA)
                    plasmas_2++;
                
                else
                    misiles_2++;
            }
            
            lasers_1=Math.min(lasers_1, lasers_2);
            plasmas_1=Math.min(plasmas_1, plasmas_2);
            misiles_1=Math.min(misiles_1, misiles_2);
            
            ArrayList<WeaponType> array_tipos = new ArrayList<WeaponType>();
            
            for(int i=0; i < (lasers_1+plasmas_1+misiles_1); i++){
                if(lasers_1 > 0){
                    array_tipos.add(WeaponType.LASER);
                    lasers_1--;
                }
                
                else if(plasmas_1 > 0){
                    array_tipos.add(WeaponType.PLASMA);
                    plasmas_1--;
                }
                
                else
                    array_tipos.add(WeaponType.MISSILE);     
            }
            
            return new Damage((ArrayList<WeaponType>) array_tipos.clone(), nShields);  
        }
    }
    
    public void discardWeapon(Weapon w){
        if(weapons.size() != 0){
            for(int i=0; i < weapons.size(); i++){
                if(weapons.get(i) == w.getType())
                    weapons.remove(i);
                    
            }
        }
        
        else if(nWeapons - 1 >= 0)
            nWeapons--;
    }
    
    public void discardShieldBooster(){
        if(nShields - 1 >= 0)
            nShields--;
    }
    
    public boolean hasNoEffect(){
        if((nShields == 0 && nWeapons==0) && weapons.isEmpty())
            return true;
        
        else
            return false;
    }
    
    public int getNShields(){
        return this.nShields;
    }
    
    public int getNWeapons(){
        return this.nWeapons;
    }
    
    public final ArrayList<WeaponType> getWeapons(){
        return this.weapons;
    }
}

File: SEGUNDO/C2/PDOO/backup/DeepSpace_ruby/pkg/DeepSpace_ruby-0.0.1/lib/damage.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
#encoding:utf-8
require_relative 'DamageToUI'
require_relative 'WeaponType'
require_relative 'Weapon'
require_relative 'ShieldBooster'


module Deepspace
class Damage
  def initialize(w,s,array)
    @nShields=s
    @nWeapons=w
    @weapons=array
  end
  
  def self.newNumericWeapons(w, s)
    new(w,s,nil)   
  end
  
  def self.newSpecificWeapons(wl,s)
    new(-1,s,wl)
  end
  
  def self.newCopy(d)
    if(d.weapons != nil)
      newSpecificWeapons(d.weapons, d.nShields)
      
    else
      newNumericWeapons(d.nWeapons,d.nShields)
    end  
  end
  
  def getUIversion
    return DamageToUI.new(self)
  end
  
  def arrayContainsType(w,t)
    i=0
    encontrado=false
    
    while i < w.length && !encontrado
      if(w.at(i).type == t)
        encontrado=true
       
      else
        i += 1
      end
    end
    
    if(encontrado)
      return (i)
    else
      return -1
    end
    
  end 
  
  def adjust(w,s)
    nShields=[@nShields,s.length].min
    
    if(@weapons == nil)
      nWeapons=[@nWeapons, w.length].min
      
      return Damage.newNumericWeapons(nWeapons, nShields)
      
    else
      laser_1=0
      plasma_1=0
      missile_1=0
      laser_2=0
      plasma_2=0
      missile_2=0
      
      i=0
      
      while i < w.length
        if(w.at(i).type == WeaponType::LASER)
          laser_1+=1
        end 
        
        if(w.at(i).type == WeaponType::PLASMA)
          plasma_1+=1
        end
        
        if(w.at(i).type == WeaponType::MISSILE)
          missile_1+=1
        end
        
        i+=1
      end
      
      i=0
      
      while i < @weapons.length
        if(@weapons.at(i) == WeaponType::LASER)
          laser_2+=1
        end 
        
        if(@weapons.at(i) == WeaponType::PLASMA)
          plasma_2+=1
        end 
        
        if(@weapons.at(i)== WeaponType::MISSILE)
          missile_2+=1
        end
        
        i+=1
      end
      
      laser_1=[laser_1,laser_2].min
      plasma_1=[plasma_1, plasma_2].min
      missile_1=[missile_1, missile_2].min 
      
      tipos_armas=Array.new
      if((laser_1+plasma_1+missile_1) <= 0)
        return Damage.newSpecificWeapons(tipos_armas,nShields)
        
      else
        i=0
        nWeapons=laser_1+plasma_1+missile_1
        
        while i < nWeapons
          if(laser_1 != 0)
            tipos_armas.push(WeaponType::LASER)
            laser_1+= -1
          elsif(plasma_1 != 0)
            tipos_armas.push(WeaponType::PLASMA)
            plasma_1 += -1
          else
            tipos_armas.push(WeaponType::MISSILE)
            
          end
          
          i+=1
        end
        
        return Damage.newSpecificWeapons(tipos_armas,nShields)
      end
    end  
  end
  
  def discardWeapon(w)
    i = 0
    
    if(@weapons != nil)
      while i < @weapons.length
        if(@weapons.at(i) == w.type)
          @weapons.delete_at(i)
        end
      
        i += 1 
      end
    
    else
      if(@nWeapons > 0)
        @nWeapons += -1
      end
    end
    
  end
  
  def discardShieldBooster
    if(@nShields > 0)
      @nShields += -1
    end
  end
  
  def hasNoEffect
    if(@weapons == nil)
      if(@nShields == 0 && @nWeapons == 0)
        return true
      
      else
        return false
      end
    
    else
      if(@nShields == 0 && @weapons.length == 0)
        return true
      
      else
        return false
      end
    end
    
  end
  
  def nShields
    @nShields
  end
  
  def nWeapons
    @nWeapons
  end
  
  def weapons
    return @weapons
  end
  
  def to_s
    salida="N escudos: #{@nShields}\nN Armas: #{@nWeapons}\n"
    
    if(@weapons != nil)
      salida = salida + @weapons.join("\n")
    end
    
    return salida
  end 
  
  private :arrayContainsType
end

end
